Add OpenACC module to core_test

gdicker1 opened the pull request:
This PR adds a new module to the test core that will be used to test OpenACC functionality and any OpenACC-specific framework routines that are added to MPAS. The only test added in this PR is based on patterns from the MPAS-A dycore. If the test core is compiled without `OPENACC=true`, the OpenACC tests are skipped.

gdicker1 commented:
Also includes changes to LDFLAGS that are in #1174.

mgduda commented:
> Also includes changes to LDFLAGS that are in #1174.

With PR #1174 merged, I think it would be safe to rebase this PR.

mgduda commented on src/core_test/mpas_test_core.F:
```diff
@@ -211,6 +212,22 @@ function test_core_run(domain) result(iErr)!{{{
+      !
+      ! Run mpas_test_openacc
+      !
+      #ifdef MPAS_OPENACC
```
Also, I think it would be good to print a blank line before this test begins. That way, in case the test before this one doesn't terminate with a blank line, we still get readable log output.

gdicker1 replied:
Gotcha, addressed in the most recent force push

mgduda commented on src/core_test/mpas_test_openacc.F:
```diff
@@ -0,0 +1,362 @@
+
+        call mpas_log_write('--- Begin OpenACC tests')
+
+        routine_name = 'openacc_test_rep_arrs'
```
Since this name gets displayed in the log file with either `PASSED` or `FAILED`, should we choose a more descriptive string to print? Maybe something like "Simple CPU-GPU reproducibility"?

gdicker1 replied:
I renamed the var to `test_log_str` and used your suggested string in the force push from [690909e ](https://github.com/MPAS-Dev/MPAS-Model/commit/690909ed0fb06b3037e997ed7828151e97f0ae2a)to [10fb55b ](https://github.com/MPAS-Dev/MPAS-Model/commit/10fb55bca892dec069413c0dca378ef2b21fcf12)

mgduda commented on src/core_test/mpas_test_core.F:
```diff
@@ -211,6 +212,22 @@ function test_core_run(domain) result(iErr)!{{{
+      !
+      ! Run mpas_test_openacc
+      !
+      #ifdef MPAS_OPENACC
```
Actually, the GNU compiler complains when the directive doesn't begin in the first column:
```
mpas_test_core.F:218:8:

  218 |       #ifdef MPAS_OPENACC
      |        1
Error: Invalid character in name at (1)
```

gdicker1 replied:
I wondered if there was a reason for it. Addressed in push from [690909e ](https://github.com/MPAS-Dev/MPAS-Model/commit/690909ed0fb06b3037e997ed7828151e97f0ae2a)to [10fb55b ](https://github.com/MPAS-Dev/MPAS-Model/commit/10fb55bca892dec069413c0dca378ef2b21fcf12)

mgduda commented on src/core_test/mpas_test_openacc.F:
```diff
@@ -0,0 +1,362 @@
+        real (kind=RKIND) :: diff
+        type (mpas_pool_type), pointer :: mesh_pool
+        integer, pointer :: nCells,nCellsSolve,cellSolveStart,cellSolveEnd
+        integer, pointer :: nEdges,nEdgesSolve,edgeSolveStart,edgeSolveEnd
```
The `cellSolveStart`, `cellSolveEnd`, `edgeSolveStart`, and `edgeSolveEnd` variables are not used and could be removed.

gdicker1 replied:
Removed in force push from [690909e ](https://github.com/MPAS-Dev/MPAS-Model/commit/690909ed0fb06b3037e997ed7828151e97f0ae2a)to [10fb55b ](https://github.com/MPAS-Dev/MPAS-Model/commit/10fb55bca892dec069413c0dca378ef2b21fcf12)

mgduda commented on src/core_test/mpas_test_openacc.F:
```diff
@@ -0,0 +1,362 @@
+           cell1 = cellsOnEdge(1,iEdge)
+           cell2 = cellsOnEdge(2,iEdge)
+
+           edge_arr_cpu(iEdge) = 0.5_RKIND * (help_arr(cell1) + help_arr(cell2))
```
It's not necessarily a problem since all cells (owned and halo) are initialized in the `help_arr` array, but note that there may be owned edges that form the boundary between an owned cell and a halo cell.

gdicker1 replied:
That crossed my mind too, so I was careful to initialize everything possible. An extension of these tests could include a GPU-direct halo exchange and some operation to "advect" those halo values.

gdicker1 commented on src/core_test/mpas_test_openacc.F:
```diff
@@ -0,0 +1,362 @@
+    !> \date   29 May 2024
+    !> \details
+    !>  Given a cpu and a gpu version of the same edge-based array, this
+    !>  this routine uses OpenACC loops to calculate the absolute difference
```
I agree, CPU seems like the best place for this (reduce complexity where possible). This probably came from a wrongly recorded note to self.

gdicker1 replied:
For the diff between the 1-d arrays on the CPU, which would be better in Fortran?

```fortran
diff = sum(abs(arr_cpu(1:nVals) - arr_gpu(1:nVals)))
```

or

```fortran
diff = sum(abs(arr_cpu(:) - arr_gpu(:)))
```

The former seems safer, but would require one more argument or a call to `size`.

mgduda replied:
Specifying the range of indices is probably safer. Note that there is a "garbage edge" at index `nEdges+1` for Registry-defined fields dimensioned by `nEdge`. I think we initialize the garbage edge, but it's probably not a guarantee. So, the difference between `arr_cpu(nEdges+1)` and `arr_gpu(nEdges+1)` could be nonzero if the garbage edge isn't initialized.

mgduda replied:
Since the `diff_arrs_1d` subroutine is essentially one line of code, I think there's a good argument for just eliminating that subroutine altogether and calculating the difference inline where we would have otherwise called `diff_arrs_1d`.

gdicker1 commented:
> With PR #1174 merged, I think it would be safe to rebase this PR.

Addressed by force push from [b5d2a99 ](https://github.com/MPAS-Dev/MPAS-Model/commit/b5d2a99751ae21422a0e8145ce3eb221cbd20f72)to [690909e ](https://github.com/MPAS-Dev/MPAS-Model/commit/690909ed0fb06b3037e997ed7828151e97f0ae2a)

gdicker1 commented on src/core_test/Registry.xml:
```diff
@@ -184,7 +184,10 @@
+	<var_struct name="openaccTest" time_levs="1">
+		<var name="edge_cpu" type="real" dimensions="nEdges"/>
+		<var name="edge_gpu" type="real" dimensions="nEdges"/>
+  </var_struct>
```
Thank you, tabs and spaces looked the same in my editor. Fixed in push from [690909e ](https://github.com/MPAS-Dev/MPAS-Model/commit/690909ed0fb06b3037e997ed7828151e97f0ae2a)to [10fb55b ](https://github.com/MPAS-Dev/MPAS-Model/commit/10fb55bca892dec069413c0dca378ef2b21fcf12)

gdicker1 commented on src/core_test/mpas_test_openacc.F:
```diff
@@ -0,0 +1,362 @@
+            ierr = ierr + 1
+        end if
+
+        call mpas_dmpar_max_int(domain % dminfo, ierr, ierr_global)
```
Yep repeated content was indeed repeated. Fixed

gdicker1 commented on src/core_test/mpas_test_openacc.F:
```diff
@@ -0,0 +1,362 @@
+        ! Deallocate any dynamic vars here
+
+        ! Make sure all threads have the max number of tests failed in
+        call mpas_dmpar_max_int(domain % dminfo, ierr, ierr_global)
```
Absolutely should have been. Fixed

gdicker1 commented on src/core_test/mpas_test_openacc.F:
```diff
@@ -0,0 +1,362 @@
+! Copyright (c) 2023 The University Corporation for Atmospheric Research (UCAR).
```
Thank you!

gdicker1 replied:
I missed it in previous rebase. Corrected in push [10fb55b ](https://github.com/MPAS-Dev/MPAS-Model/commit/10fb55bca892dec069413c0dca378ef2b21fcf12)to [cdc2ac0 ](https://github.com/MPAS-Dev/MPAS-Model/commit/cdc2ac0a29a9e85747a9c8f8bcb828af8d080e7f)

mgduda commented on src/core_test/mpas_test_openacc.F:
```diff
@@ -0,0 +1,340 @@
+        use mpas_kind_types, only : RKIND
+        use mpas_pool_routines, only : mpas_pool_get_subpool,mpas_pool_get_dimension, &
+                                       mpas_pool_get_array
+        use mpas_dmpar, only : mpas_dmpar_max_int
```
I think this `use` statement is no longer needed.

mgduda reviewed (changes requested):
I think this looks generally good. Just two minor comments to address.

gdicker1 commented:
These comments should be addressed by the push from [cdc2ac0 ](https://github.com/MPAS-Dev/MPAS-Model/commit/cdc2ac0a29a9e85747a9c8f8bcb828af8d080e7f)to [054078d ](https://github.com/MPAS-Dev/MPAS-Model/commit/054078df4c304248d7109517b26932bc0e41d626)

I removed the `diff_arrs_1d` routine, removed the `use mpas_dmpar ...` from `openacc_test_rep_arrs`, and initialized diff to 0.0 in `openacc_test_rep_arrs`

mgduda reviewed (changes requested):
The final state of the PR branch looks good to me. But, some commits (e.g., a2a82b2c) don't compile with the GNU 13.2.0 compilers:
```
mpas_test_openacc.F:343:10:

  343 |         #ifdef MPAS_OPENACC
      |          1
Error: Invalid character in name at (1)
```
If it's easier, feel free to just squash all of the commits into a single commit. Otherwise, if you'd like to preserve all commits, I think you may need to go back to the first commit on the branch to resolve the pre-processing directive issue.

gdicker1 commented:
Thank you for pointing that out, I neglected to fix the macro spacing in the first commit in the sequence. I checked with gnu on Derecho, the push from [054078d ](https://github.com/MPAS-Dev/MPAS-Model/commit/054078df4c304248d7109517b26932bc0e41d626)to [b276568 ](https://github.com/MPAS-Dev/MPAS-Model/commit/b27656892835b00cd667759407673eb8465e3459) addresses this.

mgduda reviewed (approved):
Looks good to me. Thanks!
